### PR TITLE
cannon: Implement MIPS64Memory.sol

### PR DESCRIPTION
**Pull request by Inphi**
Port MIPSMemory.sol to support 64-bit memory. This is part of the ongoing work to implement 64-bit cannon in Solidity.

Reviewer's note: MIPS64Memory.sol is similar to MIPSMemory.sol. It'll be helpful to diff both files to learn the changes made.

## Meta

part of https://github.com/ethereum-optimism/optimism/issues/12250

### CODE REVIEW

File: packages/contracts-bedrock/scripts/go-ffi/differential-testing.go [hunk 1]
```diff
@@ -366,37 +366,39 @@ func DiffTestUtils() {
 		if len(args) != 3 && len(args) != 5 && len(args) != 7 {
 			panic("Error: cannonMemoryProofWithProof requires 2, 4, or 6 arguments")
 		}
-		pc, err := strconv.ParseUint(args[1], 10, 32)
+		pc, err := strconv.ParseUint(args[1], 10, arch.WordSize)
 		checkErr(err, "Error decoding addr")
-		insn, err := strconv.ParseUint(args[2], 10, 32)
+		insn, err := strconv.ParseUint(args[2], 10, arch.WordSize)
```
Comment (mbaxter) [inline packages/contracts-bedrock/scripts/go-ffi/differential-testing.go]: (nit) We're accepting a `Word` for the`insn` param even though the instruction is always 32-bit.  Maybe we should just change the language here to label pc/insn as a generic memAddr/memValue pair.  Or else, keep `insn` 32-bit and transform insn into a Word value before setting.

File: packages/contracts-bedrock/test/cannon/MIPS64Memory.t.sol [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    /// @dev Static unit test asserting that reads reverts when a misaligned memory address is provided
+    function test_readInvalidAddress_reverts() external {
+        uint64 addr = 0x100;
+        uint64 word = 0x11_22_33_44_55_66_77_88;
+        bytes32 root;
+        bytes memory proof;
+        (root, proof) = ffi.getCannonMemory64Proof(addr, word);
+        vm.expectRevert(bytes(""));
```
Comment (mbaxter) [inline packages/contracts-bedrock/test/cannon/MIPS64Memory.t.sol]: Why don't we use revert reasons throughout?  Seems like a reason would be useful here.

**Review by mbaxter — APPROVED** [review]

File: packages/contracts-bedrock/scripts/go-ffi/differential-testing.go (hunk 1 above)
Comment (Inphi) [reply packages/contracts-bedrock/scripts/go-ffi/differential-testing.go]: good point. Will do.

File: packages/contracts-bedrock/test/cannon/MIPS64Memory.t.sol (hunk 2 above)
Comment (Inphi) [reply packages/contracts-bedrock/test/cannon/MIPS64Memory.t.sol]: it's a bit tricky to do this in yul. Lemme find a workaround.

Comment (Inphi) [reply packages/contracts-bedrock/test/cannon/MIPS64Memory.t.sol]: Fixed

File: packages/contracts-bedrock/scripts/go-ffi/differential-testing.go (hunk 1 above)
Comment (Inphi) [reply packages/contracts-bedrock/scripts/go-ffi/differential-testing.go]: Using generic memAddr/memValue for this. Also added a comment on pc/insn usage.

**Pull request merged**